crl/validator.go

```diff
@@ -20,346 +20,420 @@ package crl
+			return err
+		}
+		for _, cert := range certificates {
+			issuer := v.truststore[cert.Issuer.String()]
+			if issuer == nil {
+				// This indicates a mismatch between crlValidator truststore and tls.Config. This is a programming error.
+				return errors.New("tls.Config contains certificate from issuer that is not in the truststore: " + cert.Subject.String())
+			}
+			// add any previously unknown CRL distribution points.
+			// (Leaf cert is likely to contain CRLDistributionPoints not found in the truststore.)
+			if len(cert.CRLDistributionPoints) > 0 {
+				for _, endpoint := range cert.CRLDistributionPoints {
+					if v.crls[endpoint] == nil {
+						v.crls[endpoint] = newRevocationList(issuer)
+					}
+				}
+			}
+		}
 	}
 
-	v.listsLock.Lock()
-	defer v.listsLock.Unlock()
+	config.VerifyPeerCertificate = func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+		// TODO: the truststore should already contain ALL CAs that might be found in the verifiedChains
+		// TODO: should we check rawCerts, or verified chains?
+		// rawCerts contains raw cert data from the tls handshake.
+		// It is used together with the tls.Config to generate verifiedChains, but may contain additional certs that are not in a verified chain.
 
-	issuerName := crl.Issuer.String()
+		certificates, err := parseCertificatesSlice(rawCerts)
+		if err != nil {
+			return err
+		}
 
-	for _, cert := range crl.RevokedCertificates {
-		v.setRevoked(issuerName, cert.SerialNumber)
+		return v.validateChain(certificates)
 	}
-
-	v.lists[endpoint] = crl
-
 	return nil
 }
 
-// crlHasExpired is a replacement for pkix.CertificateList.HasExpired
-func crlHasExpired(crl *x509.RevocationList, deadline time.Time) bool {
-	return !deadline.Before(crl.NextUpdate)
+// validateChain validates that none of the certificates in the chain is revoked.
+// Certificated in chain are assumed ordered from leaf to root certificate.
+func (v *validator) validateChain(chain []*x509.Certificate) error {
+	var cert *x509.Certificate
+	var err error
+	for i := range chain {
+		cert = chain[len(chain)-1-i]
+		// check in reverse order to prevent CRL expiration errors due to revoked CAs no longer issuing CRLs
+		if err = v.validateCert(cert); err != nil {
+			return fmt.Errorf("%w: subject=%s, S/N=%s, issuer=%s", err, cert.Subject.String(), cert.SerialNumber.String(), cert.Issuer.String())
+		}
+	}
+	return nil
 }
 
-// IsSynced returns whether all the CRLs are downloaded and are not outdated (based on the offset)
-func (v *validator) IsSynced(maxOffsetDays int) bool {
-	endpoints := v.parseCRLEndpoints()
-
-	v.listsLock.RLock()
-	defer v.listsLock.RUnlock()
-
-	// Check if all CRLs have been downloaded, but if ignore CRL endpoints that are only used by expired certificates.
-	for endpoint, dependingCAs := range endpoints {
-		if !anyCertificateActive(dependingCAs) {
-			continue
+func (v *validator) validateCert(cert *x509.Certificate) error {
+	if nowFunc().After(cert.NotAfter) {
```

Comment:
I think this is already checked by golang/tls?

(otherwise, we're missing a NotBefore check..)

Reply:
It is, but `Valdiate` can be (and is) used outside the context of the standard library's tls handshake.

Reply:
changed to `nowFunc().Before(cert.NotBefore) || nowFunc().After(cert.NotAfter) ` and error to `"certificate is not valid"`